src/llvm-late-gc-lowering.cpp

```diff
@@ -1168,45 +1168,79 @@ State LateLowerGCFrame::LocalScan(Function &F) {
                             if (!(SRet->isStaticAlloca() && isa<PointerType>(ElT) && ElT->getPointerAddressSpace() == AddressSpace::Tracked)) {
                                 assert(!tracked.derived);
                                 if (tracked.all) {
                                     S.ArrayAllocas[SRet] = tracked.count * cast<ConstantInt>(SRet->getArraySize())->getZExtValue();
                                 }
                                 else {
                                     Value *arg1 = (CI->arg_begin()[1])->stripInBoundsOffsets();
-                                    AllocaInst *SRet_gc = nullptr;
-                                    if (PHINode *Phi = dyn_cast<PHINode>(arg1)) {
-                                        for (Value *V : Phi->incoming_values()) {
-                                            if (AllocaInst *Alloca = dyn_cast<AllocaInst>(V->stripInBoundsOffsets())) {
-                                                if (SRet_gc == nullptr) {
-                                                    SRet_gc = Alloca;
-                                                } else if (SRet_gc == Alloca) {
-                                                    continue;
-                                                } else {
-                                                    llvm_dump(Alloca);
-                                                    llvm_dump(SRet_gc);
-                                                    assert(false && "Allocas in Phi node should match");
-                                                }
+                                    SmallVector<AllocaInst *> gc_allocas;
+                                    SmallVector<Value*> worklist;
+                                    worklist.push_back(arg1);
+                                    while (!worklist.empty()) {
+                                        Value *V = worklist.pop_back_val();
+                                        if (AllocaInst *Alloca = dyn_cast<AllocaInst>(V->stripInBoundsOffsets())) {
+                                            gc_allocas.push_back(Alloca);
+                                        } else if (PHINode *Phi = dyn_cast<PHINode>(V)) {
+                                            for (Value *Incoming : Phi->incoming_values()) {
+                                                worklist.push_back(Incoming);
+                                            }
+                                        } else if (SelectInst *SI = dyn_cast<SelectInst>(arg1)) {
+                                            AllocaInst *TrueSRet = dyn_cast<AllocaInst>(SI->getTrueValue());
+                                            AllocaInst *FalseSRet = dyn_cast<AllocaInst>(SI->getFalseValue());
+                                            if (TrueSRet && FalseSRet) {
+                                                worklist.push_back(TrueSRet);
+                                                worklist.push_back(FalseSRet);
```

Comment:
What if `TrueSRet == FalseSRet` but it hasn't been eliminated yet?

Reply:
I think the solution is to make gc_allocas a set instead of a smallvector. So if we end up pushing the same thing twice that's still fine